File: plugins/tauri-plugin-phone-identity/android/src/main/java/io/github/biulight/phone_identity/QrFraming.kt
package io.github.biulight.phone_identity

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.node.ArrayNode
import com.fasterxml.jackson.dataformat.cbor.CBORFactory
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64

internal class EncodedQrFrame internal constructor(internal val value: String) {
    override fun toString(): String = "EncodedQrFrame([REDACTED])"
}

internal sealed class QrAssemblyStatus {
    class InProgress(val received: Int, val total: Int) : QrAssemblyStatus()

    class Complete(val message: ByteArray) : QrAssemblyStatus() {
        override fun toString(): String = "Complete(messageBytes=${message.size})"
    }
}

internal object QrFraming {
    const val DEFAULT_CHUNK_BYTES = 600
    const val MAX_MESSAGE_BYTES = 65_536
    const val MAX_FRAGMENTS = 128
    const val MAX_ASSEMBLY_AGE_MS = 30_000L
    private const val MAX_ENCODED_FRAME_CHARS = 2_048
    private const val PREFIX = "age-phone:qr1:"
    private const val VERSION = 1
    private const val FRAME_TYPE = 1
    private val digestDomain = "age-plugin-phone/qr-message-digest/v1"
        .toByteArray(Charsets.US_ASCII)
    private val cbor = ObjectMapper(CBORFactory())
    private val encoder = Base64.getUrlEncoder().withoutPadding()
    private val decoder = Base64.getUrlDecoder()

    fun isFrameCandidate(value: String): Boolean = value.startsWith(PREFIX)

    fun fragment(
        message: ByteArray,
        chunkBytes: Int = DEFAULT_CHUNK_BYTES,
        random: SecureRandom = SecureRandom(),
    ): List<EncodedQrFrame> {
        if (message.isEmpty() || message.size > MAX_MESSAGE_BYTES) throw QrException(Category.MESSAGE_SIZE)
        if (chunkBytes !in 1..DEFAULT_CHUNK_BYTES) throw QrException(Category.CHUNK_SIZE)
        val count = (message.size + chunkBytes - 1) / chunkBytes
        if (count > MAX_FRAGMENTS) throw QrException(Category.TOO_MANY_FRAGMENTS)
        val transferId = ByteArray(16).also(random::nextBytes)
        val digest = messageDigest(message)
        return (0 until count).map { index ->
            val start = index * chunkBytes
            val chunk = message.copyOfRange(start, minOf(start + chunkBytes, message.size))
            try {
                encode(
                    Frame(transferId, digest, index, count, message.size, chunk),
                )
            } finally {
                chunk.fill(0)
            }
        }
    }

    internal fun decode(encoded: String): Frame {
        if (encoded.length > MAX_ENCODED_FRAME_CHARS || !encoded.startsWith(PREFIX)) {
            throw QrException(Category.MALFORMED_FRAME)
        }
        val payload = encoded.substring(PREFIX.length)
        if (payload.isEmpty() || payload.contains('=')) throw QrException(Category.MALFORMED_FRAME)
        val encodedBytes = try {
            decoder.decode(payload)
        } catch (_: IllegalArgumentException) {
            throw QrException(Category.MALFORMED_FRAME)
        }
        if (encoder.encodeToString(encodedBytes) != payload) throw QrException(Category.MALFORMED_FRAME)
        val nodes = try {
            cbor.readTree(encodedBytes) as? ArrayNode
        } catch (_: Exception) {
            null
        } ?: throw QrException(Category.MALFORMED_FRAME)
        if (nodes.size() != 8) throw QrException(Category.MALFORMED_FRAME)
        val version = integer(nodes[0])
        if (version != VERSION) throw QrException(Category.UNSUPPORTED_VERSION)
        val type = integer(nodes[1])
        if (type != FRAME_TYPE) throw QrException(Category.UNSUPPORTED_TYPE)
        val frame = Frame(
            transferId = bytes(nodes[2], 16),
            digest = bytes(nodes[3], 32),
            index = integer(nodes[4]),
            count = integer(nodes[5]),
            totalLength = integer(nodes[6]),
            chunk = bytes(nodes[7], null),
        )
        if (frame.count !in 1..MAX_FRAGMENTS || frame.index !in 0 until frame.count ||
            frame.totalLength !in 1..MAX_MESSAGE_BYTES || frame.chunk.isEmpty() ||
            frame.chunk.size > DEFAULT_CHUNK_BYTES || frame.chunk.size > frame.totalLength ||
            !MessageDigest.isEqual(encodeCbor(frame), encodedBytes)
        ) {
            frame.chunk.fill(0)
            throw QrException(Category.MALFORMED_FRAME)
        }
        return frame
    }

    internal fun encode(frame: Frame): EncodedQrFrame {
        val value = PREFIX + encoder.encodeToString(encodeCbor(frame))
        if (value.length > MAX_ENCODED_FRAME_CHARS) throw QrException(Category.CHUNK_SIZE)
        return EncodedQrFrame(value)
    }

    private fun encodeCbor(frame: Frame): ByteArray = cbor.writeValueAsBytes(
        cbor.createArrayNode().apply {
            add(VERSION)
            add(FRAME_TYPE)
            add(frame.transferId)
            add(frame.digest)
            add(frame.index)
            add(frame.count)
            add(frame.totalLength)
            add(frame.chunk)
        },
    )

    private fun integer(node: JsonNode): Int {
        if (!node.isIntegralNumber || !node.canConvertToInt() || node.intValue() < 0) {
            throw QrException(Category.MALFORMED_FRAME)
        }
        return node.intValue()
    }

    private fun bytes(node: JsonNode, size: Int?): ByteArray {
        if (!node.isBinary) throw QrException(Category.MALFORMED_FRAME)
        val value = try {
            node.binaryValue()
        } catch (_: Exception) {
            throw QrException(Category.MALFORMED_FRAME)
        }
        if (size != null && value.size != size) throw QrException(Category.MALFORMED_FRAME)
        return value
    }

    internal fun messageDigest(message: ByteArray): ByteArray = MessageDigest.getInstance("SHA-256").run {
        update(digestDomain)
        update(0.toByte())
        digest(message)
    }

    internal class Frame(
        val transferId: ByteArray,
        val digest: ByteArray,
        val index: Int,
        val count: Int,
        val totalLength: Int,
        val chunk: ByteArray,
    ) {
        override fun toString(): String = "Frame([REDACTED])"
    }

    enum class Category {
        CHUNK_SIZE,
        CLOCK_ROLLBACK,
        CONFLICTING_FRAGMENT,
        DIFFERENT_TRANSFER,
        DIGEST_MISMATCH,
        MALFORMED_FRAME,
        MESSAGE_SIZE,
        POISONED,
        TIMEOUT,
        TOO_MANY_FRAGMENTS,
        UNSUPPORTED_TYPE,
        UNSUPPORTED_VERSION,
    }

    class QrException(val category: Category) : Exception()
}

internal class QrReassembler {
    private var active: ActiveAssembly? = null
    private var poisoned = false

    fun push(encoded: String, nowMs: Long): QrAssemblyStatus {
        if (poisoned) throw QrFraming.QrException(QrFraming.Category.POISONED)
        if (nowMs < 0) throw QrFraming.QrException(QrFraming.Category.CLOCK_ROLLBACK)
        val frame = QrFraming.decode(encoded)
        try {
            if (active == null) active = ActiveAssembly(frame, nowMs)
            val current = requireNotNull(active)
            if (nowMs < current.startedAtMs) poison(QrFraming.Category.CLOCK_ROLLBACK)
            if (nowMs - current.startedAtMs > QrFraming.MAX_ASSEMBLY_AGE_MS) {
                poison(QrFraming.Category.TIMEOUT)
            }
            if (!MessageDigest.isEqual(frame.transferId, current.transferId) ||
                !MessageDigest.isEqual(frame.digest, current.digest) ||
                frame.count != current.count || frame.totalLength != current.totalLength
            ) {
                throw QrFraming.QrException(QrFraming.Category.DIFFERENT_TRANSFER)
            }

            val existing = current.chunks[frame.index]
            if (existing != null) {
                if (!MessageDigest.isEqual(existing, frame.chunk)) {
                    poison(QrFraming.Category.CONFLICTING_FRAGMENT)
                }
            } else {
                current.receivedBytes = try {
                    Math.addExact(current.receivedBytes, frame.chunk.size)
                } catch (_: ArithmeticException) {
                    poison(QrFraming.Category.MALFORMED_FRAME)
                }
                if (current.receivedBytes > current.totalLength) {
                    poison(QrFraming.Category.MALFORMED_FRAME)
                }
                current.chunks[frame.index] = frame.chunk.copyOf()
                current.received += 1
            }

            if (current.received != current.count) {
                return QrAssemblyStatus.InProgress(current.received, current.count)
            }
            return assemble(current)
        } finally {
            frame.chunk.fill(0)
        }
    }

    fun reset() {
        clearActive()
        poisoned = false
    }

    private fun assemble(current: ActiveAssembly): QrAssemblyStatus {
        try {
            val firstSize = current.chunks.firstOrNull()?.size
                ?: poison(QrFraming.Category.MALFORMED_FRAME)
            if (firstSize == 0 || current.chunks.dropLast(1).any { it == null || it.size != firstSize } ||
                current.chunks.lastOrNull().let { it == null || it.isEmpty() || it.size > firstSize }
            ) {
                poison(QrFraming.Category.MALFORMED_FRAME)
            }
            val message = ByteArray(current.totalLength)
            var offset = 0
            for (chunk in current.chunks) {
                requireNotNull(chunk).copyInto(message, offset)
                offset += chunk.size
            }
            if (offset != message.size) {
                message.fill(0)
                poison(QrFraming.Category.MALFORMED_FRAME)
            }
            if (!MessageDigest.isEqual(QrFraming.messageDigest(message), current.digest)) {
                message.fill(0)
                poison(QrFraming.Category.DIGEST_MISMATCH)
            }
            clearActive()
            return QrAssemblyStatus.Complete(message)
        } catch (error: QrFraming.QrException) {
            throw error
        } catch (_: Exception) {
            poison(QrFraming.Category.MALFORMED_FRAME)
        }
    }

    private fun poison(category: QrFraming.Category): Nothing {
        clearActive()
        poisoned = true
        throw QrFraming.QrException(category)
    }

    private fun clearActive() {
        active?.chunks?.forEach { it?.fill(0) }
        active = null
    }

    private class ActiveAssembly(frame: QrFraming.Frame, val startedAtMs: Long) {
        val transferId = frame.transferId.copyOf()
        val digest = frame.digest.copyOf()
        val count = frame.count
        val totalLength = frame.totalLength
        val chunks = arrayOfNulls<ByteArray>(count)
        var received = 0
        var receivedBytes = 0
    }
}

File: plugins/tauri-plugin-phone-identity/ios/Core/Sources/PhoneIdentityCore/QRFraming.swift
import CryptoKit
import Foundation

package enum QRFramingError: Error {
    case chunkSize, clockRollback, conflictingFragment, differentTransfer, digestMismatch
    case malformedFrame, messageSize, poisoned, timeout, tooManyFragments, unsupportedType, unsupportedVersion

    package var category: String {
        switch self {
        case .chunkSize: return "chunk_size"
        case .clockRollback: return "clock_rollback"
        case .conflictingFragment: return "conflicting_fragment"
        case .differentTransfer: return "different_transfer"
        case .digestMismatch: return "digest_mismatch"
        case .malformedFrame: return "malformed_frame"
        case .messageSize: return "message_size"
        case .poisoned: return "poisoned"
        case .timeout: return "timeout"
        case .tooManyFragments: return "too_many_fragments"
        case .unsupportedType: return "unsupported_type"
        case .unsupportedVersion: return "unsupported_version"
        }
    }
}

package struct QRFrame {
    package let transferId: Data
    package let digest: Data
    package let index: Int
    package let count: Int
    package let totalLength: Int
    package let chunk: Data

    package init(transferId: Data, digest: Data, index: Int, count: Int, totalLength: Int, chunk: Data) {
        self.transferId = transferId; self.digest = digest; self.index = index
        self.count = count; self.totalLength = totalLength; self.chunk = chunk
    }
}

package enum QRFraming {
    package static let prefix = "age-phone:qr1:"
    package static let defaultChunkBytes = 600
    package static let maximumMessageBytes = 65_536
    package static let maximumFragments = 128
    package static let maximumAssemblyAgeMilliseconds: Int64 = 30_000
    private static let digestDomain = Data("age-plugin-phone/qr-message-digest/v1".utf8)

    package static func isCandidate(_ value: String) -> Bool { value.hasPrefix(prefix) }

    package static func fragment(_ message: Data, chunkBytes: Int = defaultChunkBytes) throws -> [String] {
        guard !message.isEmpty, message.count <= maximumMessageBytes else { throw QRFramingError.messageSize }
        guard (1...defaultChunkBytes).contains(chunkBytes) else { throw QRFramingError.chunkSize }
        let count = (message.count + chunkBytes - 1) / chunkBytes
        guard count <= maximumFragments else { throw QRFramingError.tooManyFragments }
        var transfer = Data(count: 16)
        guard transfer.withUnsafeMutableBytes({ SecRandomCopyBytes(kSecRandomDefault, 16, $0.baseAddress!) }) == errSecSuccess else {
            throw QRFramingError.malformedFrame
        }
        let digest = messageDigest(message)
        return try (0..<count).map { index in
            let start = index * chunkBytes, end = min(start + chunkBytes, message.count)
            return try encode(QRFrame(
                transferId: transfer,
                digest: digest,
                index: index,
                count: count,
                totalLength: message.count,
                chunk: Data(message[start..<end])
            ))
        }
    }

    package static func encode(_ frame: QRFrame) throws -> String {
        let data = try StrictCBOR.encode(.array([
            .unsigned(1), .unsigned(1), .bytes(frame.transferId), .bytes(frame.digest),
            .unsigned(UInt64(frame.index)), .unsigned(UInt64(frame.count)),
            .unsigned(UInt64(frame.totalLength)), .bytes(frame.chunk)
        ]))
        let value = prefix + base64URL(data)
        guard value.count <= 2_048 else { throw QRFramingError.chunkSize }
        return value
    }

    package static func decode(_ value: String) throws -> QRFrame {
        guard value.count <= 2_048, value.hasPrefix(prefix) else { throw QRFramingError.malformedFrame }
        let encoded = String(value.dropFirst(prefix.count))
        guard !encoded.isEmpty, !encoded.contains("="), let data = decodeBase64URL(encoded), base64URL(data) == encoded else {
            throw QRFramingError.malformedFrame
        }
        let values = try StrictCBOR.decode(data).exactArray(8)
        guard try values[0].exactUnsigned() == 1 else { throw QRFramingError.unsupportedVersion }
        guard try values[1].exactUnsigned() == 1 else { throw QRFramingError.unsupportedType }
        let index = Int(try values[4].exactUnsigned()), count = Int(try values[5].exactUnsigned())
        let total = Int(try values[6].exactUnsigned()), chunk = try values[7].exactBytes()
        guard (1...maximumFragments).contains(count), (0..<count).contains(index),
              (1...maximumMessageBytes).contains(total), !chunk.isEmpty,
              chunk.count <= defaultChunkBytes, chunk.count <= total else { throw QRFramingError.malformedFrame }
        return QRFrame(
            transferId: try values[2].exactBytes(16),
            digest: try values[3].exactBytes(32),
            index: index,
            count: count,
            totalLength: total,
            chunk: chunk
        )
    }

    package static func messageDigest(_ message: Data) -> Data {
        Data(SHA256.hash(data: digestDomain + Data([0]) + message))
    }

    package static func base64URL(_ data: Data) -> String {
        data.base64EncodedString().replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "=", with: "")
    }

    package static func decodeBase64URL(_ value: String) -> Data? {
        guard value.allSatisfy({ $0.isASCII && ($0.isLetter || $0.isNumber || $0 == "-" || $0 == "_") }) else { return nil }
        var standard = value.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        standard += String(repeating: "=", count: (4 - standard.count % 4) % 4)
        return Data(base64Encoded: standard)
    }
}

package enum QRAssemblyStatus { case inProgress(received: Int, total: Int), complete(Data) }

package final class QRReassembler {
    private var active: Active?
    private var poisoned = false

    package init() {}

    package func push(_ encoded: String, nowMilliseconds: Int64) throws -> QRAssemblyStatus {
        guard !poisoned else { throw QRFramingError.poisoned }
        guard nowMilliseconds >= 0 else { throw QRFramingError.clockRollback }
        let frame = try QRFraming.decode(encoded)
        if active == nil { active = Active(frame: frame, started: nowMilliseconds) }
        guard let current = active else { throw QRFramingError.malformedFrame }
        if nowMilliseconds < current.started { try poison(.clockRollback) }
        if nowMilliseconds - current.started > QRFraming.maximumAssemblyAgeMilliseconds { try poison(.timeout) }
        guard frame.transferId == current.transferId, frame.digest == current.digest,
              frame.count == current.count, frame.totalLength == current.totalLength else {
            throw QRFramingError.differentTransfer
        }
        if let existing = current.chunks[frame.index] {
            if existing != frame.chunk { try poison(.conflictingFragment) }
        } else {
            current.receivedBytes += frame.chunk.count
            if current.receivedBytes > current.totalLength { try poison(.malformedFrame) }
            current.chunks[frame.index] = frame.chunk
            current.received += 1
        }
        guard current.received == current.count else { return .inProgress(received: current.received, total: current.count) }
        guard let first = current.chunks.first ?? nil, !first.isEmpty,
              current.chunks.dropLast().allSatisfy({ $0?.count == first.count }),
              let last = current.chunks.last ?? nil, !last.isEmpty, last.count <= first.count else {
            try poison(.malformedFrame)
        }
        let message = current.chunks.compactMap { $0 }.reduce(into: Data()) { $0.append($1) }
        guard message.count == current.totalLength else { try poison(.malformedFrame) }
        guard QRFraming.messageDigest(message) == current.digest else { try poison(.digestMismatch) }
        active = nil
        return .complete(message)
    }

    package func reset() { active = nil; poisoned = false }

    private func poison(_ error: QRFramingError) throws -> Never {
        active = nil; poisoned = true; throw error
    }

    private final class Active {
        let transferId: Data, digest: Data, count: Int, totalLength: Int, started: Int64
        var chunks: [Data?], received = 0, receivedBytes = 0
        init(frame: QRFrame, started: Int64) {
            transferId = frame.transferId; digest = frame.digest; count = frame.count
            totalLength = frame.totalLength; self.started = started
            chunks = Array(repeating: nil, count: frame.count)
        }
    }
}
